Remove old ES nodes PersistentVolumeClaims

sebgl opened the pull request:
Let's not keep PVCs around when Elasticsearch nodes are removed, since
it can be very confusing if there are reused in a different context (eg.
new cluster with the same old name, or upscale long after a downscale).

This is done in 2 ways:
- set the ownerRef of Claims in the StatefulSet resource to the ES resource, to delete PVCs on cluster deletion.
- delete unused PVCs, based on their naming conventions and what we
expect to exist from actual & expected StatefulSets.

Note we do not remove PVs, only PVCs. PVs deletion is left to the
storage class reclaim policy responsibility.

Fixes https://github.com/elastic/cloud-on-k8s/issues/1288.

sebgl commented on pkg/controller/elasticsearch/driver/pvc.go:
```diff
@@ -0,0 +1,122 @@
+// reconcilePVCsOwnerRef ensures PVCs created for this Elasticsearch cluster have an owner ref set to
+// the Elasticsearch resource, so they are deleted automatically upon Elasticsearch deletion.
+// A subtle race condition exists: users may still end up with leftover PVCs if the Elasticsearch resource
+// gets deleted right after creation or update, before this is called.
```
To handle this we could run a gc function outside the scope of a single cluster reconciliation (ie. working with all clusters), removing resources not related to any existing cluster.

I think the chance that race condition occurs is rather low (update the ES resource so we get new PVCs created, then immediately stop or disconnect the operator before it gets a chance to patch newly created PVCs). So may not be worth optimising for. And even if it may happen, then it's probably not a big deal. Thoughts?

thbkrkr replied:
I also think it's unlikely but I'm constantly reproducing the race with: `k apply -f es.yml; sleep 1; k delete es --all`. By reapplying the ES resource, the PV are reused.

sebgl replied:
@anyasabo [suggestion](https://github.com/elastic/cloud-on-k8s/issues/1288) to set the PVC ownerRef within the claim template in the StatefulSet spec should allow us to work around that. No need to reconcile owner ref after the fact, PVCs get created using owner ref specified in the StatefulSet spec at creation time.

sarjeet2013 commented:
Overall looks good, but one caveat with this is:

1. ownerReference is set on ES object 
2. This is not configurable to enable/disable it, this require documentation to let users aware of this behavior.

Regarding 1)
a) all related pvcs will live as long as ES resource isn't deleted. 
b) There will be race condition (as you mentioned) when a ES resource is created and deleted before the PVC cleaner reconciliation is run.

In case of a) Can the ownerReference be set for statefulsets, instead of ES itself? It will take care of cleaning PVCs per nodeGroup/StatefulSets and can have less PVCs lying around for the lifetime of ES cluster.

In case of b) I have tried an admission webhook on PVCs "Create" and it works fine and won't run into the race-condition even though the ES cluster is created and deleted intermittently. Can this be a possible solution?

Regarding 2) maybe a doc jira to track it?

david-kow commented on pkg/controller/elasticsearch/driver/pvc.go:
```diff
@@ -0,0 +1,122 @@
+	expectedPVCs := append(actualStatefulSets.PVCNames(), expectedStatefulSets.PVCNames()...)
+	var toRemove []corev1.PersistentVolumeClaim
+	for _, pvc := range pvcs {
+		if !stringsutil.StringInSlice(pvc.Name, expectedPVCs) {
```
Use map (set) instead?

sebgl replied:
That's a small optimization given the relatively small (<1k items) size of our arrays here, but yeah, let's do it! 👍

david-kow commented on pkg/controller/elasticsearch/driver/pvc.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+// reconcilePVCsOwnerRef ensures PVCs created for this Elasticsearch cluster have an owner ref set to
+// the Elasticsearch resource, so they are deleted automatically upon Elasticsearch deletion.
```
As it was noted already, in some scenarios ES deletion is not a part of the lifecycle, ie. cluster could be scaled up/down, but never deleted. In such a case, which I think is not uncommon, this change won't help.

sebgl replied:
Not sure I understand your comment; I think this is covered already?

david-kow replied:
I wanted to say that cleaning up after ES resource is deleted might not help in many cases, but yes, it's already covered :)

sebgl commented:
Last commit changes the logic a little bit:

We now set an ownerReference to the (existing) Elasticsearch object, in the PVClaim template, inside the StatefulSet spec. As a result, PVCs get created automatically with the ownerReference already set; we don't have to reconcile the ownerReference after the fact, and we avoid the race condition mentioned above.

@sarjeet2013 thanks for your feedback! A few comments below:

> Can the ownerReference be set for statefulsets, instead of ES itself?

Technically, yes, but the trick above that sets the ownerRef to the ES resource when building the StatefulSet allows us to avoid the race condition mentioned above. 

> It will take care of cleaning PVCs per nodeGroup/StatefulSets and can have less PVCs lying around for the lifetime of ES cluster.

The purpose of the `GarbageCollectPVCs()` function in this PR is to remove any useless PVC that may exist. This covers 2 cases:

1. the StatefulSet was removed
2. the StatefulSet was scaled down (eg. from 30 replicas to 10 replicas)

Case 1) would be done automatically if we switched the ownerRef to the StatefulSet instead of the ES resource, but we would have to do 2) anyway (using the same code, basically).

> In case of b) I have tried an admission webhook on PVCs "Create" and it works fine and won't run into the race-condition even though the ES cluster is created and deleted intermittently. Can this be a possible solution?

Could be yes, but I think the current approach is better. Also we have seen many users reporting problems with webhooks (need to whitelist apiserver -> webhook connection in their k8s environment), which makes it less than ideal. In general I think we should consider webhooks as optional nice-to-have.

> Regarding 2) maybe a doc jira to track it?

Indeed, this definitely needs to be documented. I added a comment in https://github.com/elastic/cloud-on-k8s/issues/1734.

sarjeet2013 commented:
@sebgl Agree, Current approach is much better and takes care of race-condition scenario fine. 

LGTM

thbkrkr commented on pkg/controller/elasticsearch/driver/pvc.go:
```diff
@@ -0,0 +1,65 @@
+	// consider Pods in the process of being deleted (but not deleted yet), since already covered
+	// by checking expectations earlier in the process.
+	// Then, just return existing PVCs that are not part of that list.
+	expectedPVCs := stringsutil.SliceToMap(append(actualStatefulSets.PVCNames(), expectedStatefulSets.PVCNames()...))
```
Nit variable naming: that's not really the `expectedPVCs` given it's the merge of the actual and the expected ssets. It's more `pvcsToKeep` or just `toKeep` to be aligned with `toRemove`.
